lib/propolis/src/hw/qemu/pvpanic.rs

```diff
@@ -0,0 +1,80 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::sync::{
+    atomic::{AtomicUsize, Ordering::Relaxed},
+    Arc,
+};
+
+use crate::common::*;
+use crate::pio::{PioBus, PioFn};
+
+pub struct QemuPioPvpanic {
+    counts: Arc<PanicCounts>,
+}
+
+#[derive(Debug, Default)]
+pub struct PanicCounts {
+    host_handled: AtomicUsize,
+    guest_handled: AtomicUsize,
```

Comment:
From a design standpoint, I wonder if we should just keep the counts in a Mutex, since we're not expecting high event rates or contention.


The bigger question (in my mind) being how do we want stats consumers to reach in for the data?:
1. Holding an Arc to the data bits like you have now
2. Keeping and Arc to the device itself, and proving a method on the device to emit a current snapshot of the data.

My inclination would be the latter, but it's not 100% clear.

Reply:
> From a design standpoint, I wonder if we should just keep the counts in a Mutex, since we're not expecting high event rates or contention.

I guess it's a question of whether we consider a mutex to be more or less conceptual complexity than an atomic, if performance isn't an issue here. 

My preference is normally to avoid reaching for a `Mutex` in situations where an atomic is sufficient, regardless of whether or not the data is contended enough for there to be a performance impact. The atomic counter feels conceptually simpler to me: incrementing the counter is just a one-instruction write to an integer, rather than requiring an allocation and syscalls for `pthread_mutex`...

If there's a strong preference for a mutex, though, I'm happy to change this. 



>   1. Holding an Arc to the data bits like you have now
> 
>   2. Keeping and Arc to the device itself, and proving a method on the device to emit a current snapshot of the data.
> 
> 
> My inclination would be the latter, but it's not 100% clear.

Hmm, I think you're right that it's simpler to just have a single `Arc`ed type for the device itself, and just having the Oximeter `Producer` hold a reference to the `Arc`ed device itself. The reason I didn't do that was that I thought, eventually, we would also want to support the PCI bus PVPANIC device, and we would want panics reported using the PCI device to increment the same counters as the ISA bus device, so they would need to share the counters.

Looking at it again, though, in light of your comment https://github.com/oxidecomputer/propolis/pull/596#discussion_r1446649693, I suppose we could have one `Pvpanic` struct that's potentially registered as both a PCI device and a PIO device, so we don't actually need a separate type for each device.

Reply:
> I guess it's a question of whether we consider a mutex to be more or less conceptual complexity than an atomic, if performance isn't an issue here.

Atomics makes me at least acknowledge that memory ordering semantics are a thing (which I remain not especially confident about), even if I'm not using them in a context where such ordering will have a meaningful impact.  For that reason (and a few more), I prefer `Mutex` over atomics unless there's a pressing reason (performance or otherwise).  Most of the time, I just want consistent access to all the fields covered by one.

> I suppose we could have one Pvpanic struct that's potentially registered as both a PCI device and a PIO device.

If we were to expose it via both interfaces, doing so with a single set of counters (be it via one device or two) definitely seems preferable.